--- Makefile	2006-12-02 18:52:05.000000000 -0500
+++ Makefile1	2006-12-02 18:53:26.000000000 -0500
@@ -38,7 +38,7 @@
 	$(CC) $(CFLAGS) $(LDFLAGS) -L.libs -lseom -lX11 -lGL -o example $<
 
 $(APPS): libseom.la
-	$(CC) $(CFLAGS) $(LDFLAGS) -L.libs -lseom $($@LIBS) -o $@ src/$@/main.c
+	$(CC) $(CFLAGS) $(LDFLAGS) -L.libs -o $@ src/$@/main.c -lseom $($@LIBS)
 
 install: libseom.la $(APPS)
 	install -m 0755 -d $(DESTDIR)/include/seom $(DESTDIR)/$(LIBDIR) $(DESTDIR)/bin
