diff -purN kdelibs-3.5.3.orig/kdecore/kicontheme.cpp kdelibs-3.5.3.tango/kdecore/kicontheme.cpp
--- kdelibs-3.5.3.orig/kdecore/kicontheme.cpp	2005-09-10 03:27:12.000000000 -0500
+++ kdelibs-3.5.3.tango/kdecore/kicontheme.cpp	2006-06-04 11:04:21.000000000 -0500
@@ -517,6 +517,20 @@ KIconThemeDir::KIconThemeDir(const QStri
         mContext = KIcon::Application;
     else if (tmp == "Actions")
         mContext = KIcon::Action;
+    else if (tmp == "Animations")
+        mContext = KIcon::Animation;
+    else if (tmp == "Categories")
+        mContext = KIcon::Category;
+    else if (tmp == "Emblems")
+        mContext = KIcon::Emblem;
+    else if (tmp == "Emotes")
+        mContext = KIcon::Emote;
+    else if (tmp == "International")
+        mContext = KIcon::International;
+    else if (tmp == "Places")
+        mContext = KIcon::Place;
+    else if (tmp == "Status")
+        mContext = KIcon::StatusIcon;
     else {
         kdDebug(264) << "Invalid Context= line for icon theme: " << mDir << "\n";
         return;
diff -purN kdelibs-3.5.3.orig/kdecore/kicontheme.h kdelibs-3.5.3.tango/kdecore/kicontheme.h
--- kdelibs-3.5.3.orig/kdecore/kicontheme.h	2005-09-10 03:27:12.000000000 -0500
+++ kdelibs-3.5.3.tango/kdecore/kicontheme.h	2006-06-04 11:04:21.000000000 -0500
@@ -51,7 +51,14 @@ public:
       Application, ///< An icon that represents an application.
       Device, ///< An icon that represents a device.
       FileSystem, ///< An icon that represents a file system.
-      MimeType ///< An icon that represents a mime type (or file type).
+      MimeType, ///< An icon that represents a mime type (or file type).
+      Animation, ///< An icon that is animated.
+      Category, ///< An icon that represents a category.
+      Emblem, ///< An icon that adds information to an existing icon.
+      Emote, ///< An icon that expresses an emotion.
+      International, ///< An icon that represents a country's flag.
+      Place, ///< An icon that represents a location (e.g. 'home', 'trash').
+      StatusIcon ///< An icon that represents an event.
     };
 
     /**
diff -purN kdelibs-3.5.3.orig/kio/kfile/kicondialog.cpp kdelibs-3.5.3.tango/kio/kfile/kicondialog.cpp
--- kdelibs-3.5.3.orig/kio/kfile/kicondialog.cpp	2005-10-10 10:05:43.000000000 -0500
+++ kdelibs-3.5.3.tango/kio/kfile/kicondialog.cpp	2006-06-04 11:04:21.000000000 -0500
@@ -331,6 +331,13 @@ void KIconDialog::init()
     mpCombo->insertItem(i18n("Devices"));
     mpCombo->insertItem(i18n("Filesystems"));
     mpCombo->insertItem(i18n("Mimetypes"));
+    mpCombo->insertItem(i18n("Animations"));
+    mpCombo->insertItem(i18n("Categories"));
+    mpCombo->insertItem(i18n("Emblems"));
+    mpCombo->insertItem(i18n("Emotes"));
+    mpCombo->insertItem(i18n("International"));
+    mpCombo->insertItem(i18n("Places"));
+    mpCombo->insertItem(i18n("Status"));
     mpCombo->setFixedSize(mpCombo->sizeHint());
     mpBrowseBut->setFixedWidth(mpCombo->width());
 
