Index: gnome-main-menu/control-center/src/control-center.c
===================================================================
--- gnome-main-menu.orig/control-center/src/control-center.c
+++ gnome-main-menu/control-center/src/control-center.c
@@ -139,7 +139,7 @@ main (int argc, char *argv[])
 	}
 
 	AppShellData *app_data = appshelldata_new (
-		"preferences.menu", NULL, CONTROL_CENTER_PREFIX, GTK_ICON_SIZE_DIALOG);
+		"settings.menu", NULL, CONTROL_CENTER_PREFIX, GTK_ICON_SIZE_DIALOG);
 	generate_categories (app_data);
 
 	GSList *actions = get_actions_list ();
